debian/samsparser.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH sams 1 "January 25, 2008"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
samsparser \- squid access.log parser
.SH SYNOPSIS
.B samsparser
.RI [ options ] " files" ...
.br
.B bar
.RI [ options ] " files" ...
.SH DESCRIPTION
This manual page documents briefly the
.B samsparser
command.
.PP
.\" TeX users may be more comfortable with the \fB<whatever>\fP and
.\" \fI<whatever>\fP escape sequences to invode bold face and italics, 
.\" respectively.
\fBsamsparser\fP is a program that...
.SH OPTIONS
These programs follow the usual GNU command line syntax, with long
options starting with two dashes (`-').
A summary of options is included below.
For a complete description, see the Info files.
.TP
.B \-h, \-\-help
Show summary of options.
.TP
.B \-v, \-\-version
Show version of program.
.SH SEE ALSO
.BR samsdaemon (1),
.BR sams2redir (1).
.br
.SH AUTHOR
samsparser was written by SAMS Team.
.PP
This manual page was written by Pavel Vinogradov <Pavel.Vinogradov@nixdev.net>,
for the Debian project (but may be used by others).


NEWS
История версий БД
   База данных имеет свою версию, которая хранится в websettings.s_version
   а) БД считается "актуальной" если первые 3 числа ее версии сопадают с теми-же числами версии пакета.
      Это означает что приложения работают в полнофункциональном режиме.
   б) БД считается "допустимой" если первое и второе число сопадают с теми-же числами версии пакета.
      Это означает что приложения могут работать, но некоторые функции будут недоступны.
   в) БД считается "тестовой", если второе и третье число равны 9.
      Может быть использовано в целях разработки, тестов и т.д.

   1.9.9 - первая тестовая версия
   2.9.9 - добавлено применение модулей sysinfo
           CREATE TABLE `sysinfo` (
             `s_proxy_id` INT              NOT NULL ,
             `s_name`     VARCHAR( 50 )    NOT NULL ,
             `s_version`  VARCHAR( 10 )    NOT NULL ,
             `s_author`   VARCHAR( 30 )        NULL DEFAULT 'anonymous',
             `s_info`     VARCHAR( 1024 )  NOT NULL DEFAULT 'not available',
             `s_date`     DATETIME         NOT NULL ,
             `s_status`   INT              NOT NULL
           ) ENGINE = MYISAM;
           UPDATE websettings SET s_version='2.9.9';
   3.9.9 - добавлена функция перенаправления
           ALTER TABLE `redirect` ADD `s_dest` VARCHAR( 128 ) NULL ;
   4.9.9 - добавлена функция перевода пользователя в другой шаблон при превышении лимита
           ALTER TABLE `shablon` ADD `s_shablon_id2` BIGINT( 20 ) UNSIGNED NULL ;

081116
  Вторичный шаблон пользователя, баннерорезка, добавление комментариев,
  оптимизация работы с группами url адресов
  багфиксы: 
  1) в движке MySQL если числовое значение в БД равно NULL, вылетали демоны
  2) при перечитывании настроек список плагинов системной информации не очищался
081125
  Отправка письма администратору при блокировке пользователя (или переводе во временный шаблон)
081128
  Принимаются во внимание 2 параметра прокси сервера: регистр букв в домене и имени пользователя.
081129
  1) Перевод описания таблиц из формата doc в формат html с частичной проверкой актуальности.
  2) Документация по исходникам собирается и устанавливается, если найден doxygen.
     Скорей всего rpm после этого компилироваться не будет.
081130
  Начало написания документации по установке, настройке и использованию.
081203
  Исправление ошибки выполнения make install на FreeBSD (find: unknown option -iwholename)
  Небольшое дополнение документации
  Документация по исходникам теперь не собирается по умолчанию, для сборки необходимо указать
    опцию --with-srcdocs во время configure
081205
  Учитывается proxy.s_usedomain при создании acl, поиске и создании пользователей.
  Исправлена ошибка при создании БД. Не учитывалось что изменилась таблица shablon.
  Дополнение документации.
081207
  При автоматическом создании пользователя, он не вносился в squid.conf до следующего реконфигурирования.
  Не создавалась таблица sysinfo.
  Обновлена документация.
  Исправлены мелкие ошибки при создании пакета для RedHat.
081209
  Удаление линка wbinfo из репозитария, т.к. более не используется и вносит смуту при построении пакетов.
  Параметры proxy.s_bigd, proxy.s_bigu и proxy.s_squidbase теперь корректно отображаются в web и обновляются в БД.
081217
  Добавлена информация, выводимая на экран, в конце этапа configure.
  Корректная работа команды make uninstall
  Обновление документации (раздел требований)
  Fixed #78
081220
  Исправлена и расширена настройка авторизации OpenLDAP и загрузка пользователей. Требует доработки.
090328
  Доделан механизм регистрации/разрегистрации и автоматического удаления классов типа DBquery.
090329
  Счетчики пользователей после окончания периода ограничения очищаются только если в настройках прокси
  установлен флаг "Автоматически очищать счетчики трафика пользователей"
090407
  В samsdaemon и samsredir учитывается уровень отладочных сообщений, установленный в БД, но только если он
  не был явно задан из командной строки. При нестандартном периоде ограничения трафика в шаблоне, обновляется
  дата следующей очистки, но пока только в оперативной памяти (в БД не изменяется). Незначительное обновление
  документации.
090418
  Иправлена ошибка в определении IP адреса по доменному имени.
  Исправлена несвоевременная разрегистрация MYSQLQuery.
  Добавлено использование внутреннего кеша DNS ответов
090419
  Дата следующей очистки счетчиков у шаблонов с нестандартным периодом теперь обновляется в БД.
  Исправлена регистрация/разрегистрация PgQuery.
  Обновлена документация исходных кодов.
090425
  Обновлена документация исходных кодов. Добавлено описание отличий версии 1 и 2.
090428
  Оптимизирован поиск пользователей. Отключен кеш DNS ответов (необходимо использование TTL ответов)
090429
  Немного подправлено построение rpm пакетов.
090502
  Исправлено отображение статуса пользователя и ограничение трафика в веб интерфейсе.
  При удалении временного интервала в шаблоне теперь не переходит на редактирование другого шаблона.
  Правильно устанавливается значение "Запрет доступа ко всем url" при редактировании шаблона.
090503
  Период лимита шаблона 1 день.
  Правильно устанавливается значение уровня отладочных сообщений при редактировании прокси.
  Добавлено управление системными плагинами. Чтобы заработало, необходимо модифицировать БД.
  TRUNCATE TABLE sysinfo;
  ALTER TABLE  sysinfo ADD s_row_id SERIAL PRIMARY KEY FIRST;
090504
  Добавлена поддержка ограничения по расширениям файлов (пока только для редиректора)
  Теперь в разделы acl и http_access в squid.conf ничего не пишется, если используется встроенный редиректор.
090506
  Исправлена LDAP авторизация пользователя в веб интерфейсе
090507
  Исправлен импорт пользователей из LDAP
  Исправлено редактирование списков регулярных выражений.
  Теперь принимается и частичное совпадение url с регулярным выражением
090511
  Значительно переделан интерфейс редактирования настроек прокси сервера.

trunk/src/samsparser.cpp
/***************************************************************************
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/
#include <unistd.h>
#include <getopt.h>
#include <stdlib.h>

#include "config.h"
#include "configmake.h"

#include "debug.h"
#include "samsconfig.h"
#include "squidlogparser.h"
#include "dbcleaner.h"
#include "datefilter.h"
#include "userfilter.h"
#include "processmanager.h"
#include "localnetworks.h"
#include "samsuserlist.h"
#include "templatelist.h"
#include "proxy.h"

/**
 *  Выводит список опций командной строки с кратким описанием
 */
void usage ()
{
  cout << endl;
  cout << "NAME" << endl;
  cout << "    samsparser - parse squid log file and update database." << endl;
  cout << endl;
  cout << "SYNOPSIS" << endl;
  cout << "    samsparser [COMMAND] [OPTION]..." << endl;
  cout << endl;
  cout << "COMMANDS" << endl;
  cout << "    Some commands may be used at one runtime." << endl;
  cout << endl;
  cout << "    -h, --help" << endl;
  cout << "                Show this help screen and exit." << endl;
  cout << "    -V, --version" << endl;
  cout << "                Print program version number and exit." << endl;
  cout << "    -u, --update" << endl;
  cout << "                Update database (DEFAULT)." << endl;
  cout << "                If --file option is present, this file is used for input, instead of real log file." << endl;
  cout << "    -e, --export" << endl;
  cout << "                Export cache from a database to a file." << endl;
  cout << "                If --file option is not present, squid.log is used for output." << endl;
  cout << "    -c, --clear" << endl;
  cout << "                Clear user's counters." << endl;
  cout << "    -t, --truncate" << endl;
  cout << "                Truncate cache in database." << endl;
  cout << endl;
  cout << "Mandatory arguments to long options are mandatory for short options too." << endl;
  cout << "The following options are available:" << endl;
  cout << endl;
  cout << "OPTIONS" << endl;
  cout << "    -f, --file=FILE" << endl;
  cout << "                Use FILE for update or export (DEFAULT: squid.log)." << endl;
  cout << "    -U, --user=LIST" << endl;
  cout << "                Perform all operations for listed users only (DEFAULT: all)." << endl;
  cout << "                LIST is a list of users, separated by comma (e.g. DOMAIN+user1,192.168.1.1,user2)." << endl;
  cout << "    -D, --date=DATE_INTERVAL" << endl;
  cout << "                Perform all operations for listed date interval only (DEFAULT: all)." << endl;
  cout << "                DATE_INTERVAL is two dates, separated by comma." << endl;
  cout << "                The date specification is YYYY-MM-DD" << endl;
  cout << "                E.g. -D 2007-01-01,2007-01-31." << endl;
  cout << "    -v, --verbose" << endl;
  cout << "                Produce more output." << endl;
  cout << "    -d, --debug=LEVEL" << endl;
  cout << "                Produce lots of debugging information depend on LEVEL." << endl;
  cout << "                For more information about possible LEVEL values refer to a developer." << endl;
  cout << "    -l, --logger=LOGGER" << endl;
  cout << "                Set engine for messages output. Possible values for LOGGER are: console, syslog, file." << endl;
  cout << "                In case of file you can set filename for output (DEFAULT: samsparser.log)." << endl;
  cout << "                E.g. -l syslog" << endl;
  cout << "                E.g. -l file:/path/to/file" << endl;
  cout << "    -C, --config=FILE" << endl;
  cout << "                Use config file FILE." << endl;
  cout << "    -w, --wait-myself" << endl;
  cout << "                If already running program found, do not exit immediatly," << endl;
  cout << "                but wait until it ends and finish the task." << endl;
  cout << endl;
  cout << "DESCRIPTION" << endl;
  cout << "    If no one command is listed, program updates database (--update)." << endl;
  cout << "    If more then one command is listed, program performs all operations in the following order:" << endl;
  cout << "      1) Update database (--update command)." << endl;
  cout << "      2) Export cache to a file (--export command)." << endl;
  cout << "      3) Clear user's counters (--clear command)." << endl;
  cout << "      4) Truncate cache in database (--truncate command)." << endl;
  cout << "    For all operations the filters (--user and --date) are applied, if any." << endl;
  cout << endl;
}

/**
 *  Выводит версию программы и немного информации
 */
void version ()
{
  cout << "samsparser " << VERSION << endl;
  cout << "Written by anonymous." << endl;
  cout << endl;
  cout << "This program comes with NO WARRANTY; not even for MERCHANTABILITY" << endl;
  cout << "or FITNESS FOR A PARTICULAR PURPOSE." << endl;
}

#define CMD_NONE      0x0
#define CMD_UPDATE    0x1
#define CMD_EXPORT    0x2
#define CMD_CLEAR     0x4
#define CMD_TRUNCATE  0x8

/** @todo Выбирать путь к pid файлу из опций configure
*/
int main (int argc, char *argv[])
{
  int commands;
  int parse_errors = 0;
  int c;
  int err;
  uint dbglevel = 0;
  bool wait_myself = false;
  string optname = "";
  string logfilename = "";
  DateFilter *dateFilter = NULL;
  UserFilter *userFilter = NULL;
  bool verbose = false;
  string log_engine = "";
  string config_file = SYSCONFDIR;
  config_file += "/sams2.conf";

  static struct option long_options[] = {
    {"help",        0, 0, 'h'},    // Показывает справку по опциям командной строки и завершает работу
    {"version",     0, 0, 'V'},    // Показывает версию программы и завершает работу
    {"update",      0, 0, 'u'},    // Обновляет кэш в БД и счетчики пользователей
    {"export",      0, 0, 'e'},    // Экспортирует кэш в файл
    {"clear",       0, 0, 'c'},    // Очищает счетчики пользователей
    {"truncate",    0, 0, 't'},    // Очищает кэш в БД
    {"file",        1, 0, 'f'},    // Имя файла для импорта или экспорта
    {"user",        1, 0, 'U'},    // Фильтр по пользователям
    {"date",        1, 0, 'D'},    // Фильтр по датам
    {"verbose",     0, 0, 'v'},    // Устанавливает режим многословности
    {"debug",       1, 0, 'd'},    // Устанавливает уровень отладки
    {"logger",      1, 0, 'l'},    // Устанавливает движок вывода сообщений
    {"config",      1, 0, 'C'},    // Использовать альтернативный конфигурационный файл
    {"wait-myself", 0, 0, 'w'},    // Устанавливает движок вывода сообщений
    {0, 0, 0, 0}
  };

  commands = CMD_NONE;
  while (1)
    {
      int option_index = 0;

      c = getopt_long (argc, argv, "hVuectf:U:D:vd:l:C:w", long_options, &option_index);
      if (c == -1)              // no more options
        break;
      switch (c)
        {
        case 0:
          optname = long_options[option_index].name;
          break;
        case 'h':
          usage ();
          exit (0);
          break;
        case 'V':
          version ();
          exit (0);
          break;
        case 'u':
          commands += CMD_UPDATE;
          break;
        case 'e':
          commands += CMD_EXPORT;
          break;
        case 'c':
          commands += CMD_CLEAR;
          break;
        case 't':
          commands += CMD_TRUNCATE;
          break;
        case 'f':
          logfilename = optarg;
          break;
        case 'U':
          WARNING ("Filter for users is NOT implemented.");
          userFilter = new UserFilter (optarg);
          break;
        case 'D':
          dateFilter = new DateFilter (optarg);
          break;
        case 'v':
          verbose = true;
          break;
        case 'd':
          if (sscanf (optarg, "%d", &dbglevel) != 1)
            dbglevel = 0;
          break;
        case 'l':
          log_engine = optarg;
          break;
        case 'C':
          config_file = optarg;
          break;
        case 'w':
          wait_myself = true;
          break;
        case '?':
          break;
        default:
          printf ("?? getopt return character code 0%o ??\n", c);
        }
    }

  Logger::setSender("samsparser");
  SamsConfig::useFile (config_file);

  // Сначала прочитаем конфигурацию, параметры командной строки
  // имеют приоритет, потому анализируются позже
  SamsConfig::reload ();

  Logger::setEngine (log_engine);
  Logger::setVerbose (verbose);
  Logger::setDebugLevel (dbglevel);

  if (parse_errors > 0)
    {
      usage ();
      exit (parse_errors);
    }

  string squidlogdir = SamsConfig::getString (defSQUIDLOGDIR, err);
  string squidcachefile = SamsConfig::getString (defSQUIDCACHEFILE, err);

  if (squidlogdir.empty () || squidcachefile.empty ())
    {
      ERROR ("Either " << defSQUIDLOGDIR << " or " << defSQUIDCACHEFILE << " not defined. Check config file.");
      exit (1);
    }

  int proxyid = SamsConfig::getInt (defPROXYID, err);
  if (err != ERR_OK)
    {
      ERROR ("No proxyid defined. Check " << defPROXYID << " in config file.");
      exit (1);
    }

  ProcessManager process;

  if (!process.start ("samsparser", wait_myself))
    {
      exit (2);
    }

  SquidLogParser *parser = NULL;
  DBCleaner *cleaner = NULL;

  if ((commands & CMD_CLEAR) == CMD_CLEAR || (commands & CMD_TRUNCATE) == CMD_TRUNCATE)
    {
      cleaner = new DBCleaner ();
      cleaner->setDateFilter (dateFilter);
      cleaner->setUserFilter (userFilter);
    }

  if (commands == CMD_NONE)
    commands += CMD_UPDATE;

  bool parseFromBegin = true;
  while (commands != CMD_NONE)
    {
      if ((commands & CMD_UPDATE) == CMD_UPDATE)
        {
          INFO ("+++ Updating database");
          if (logfilename.empty ())
            {
              logfilename = squidlogdir + "/" + squidcachefile;
              parseFromBegin = false;
            }

          parser = new SquidLogParser (proxyid);

          parser->setDateFilter (dateFilter);
          parser->setUserFilter (userFilter);
          parser->parseFile (logfilename, parseFromBegin);

          delete parser;

          commands -= CMD_UPDATE;
        }
      else if ((commands & CMD_EXPORT) == CMD_EXPORT)
        {
          INFO ("+++ Exporting to the file");
          if (logfilename.empty ())
            logfilename = "squid.log";

          commands -= CMD_EXPORT;
        }
      else if ((commands & CMD_CLEAR) == CMD_CLEAR)
        {
          INFO ("+++ Clearing user's counters");
          cleaner->clearCounters ();
          commands -= CMD_CLEAR;
        }
      else if ((commands & CMD_TRUNCATE) == CMD_TRUNCATE)
        {
          INFO ("+++ Truncating cache in the database");
          cleaner->clearCache ();
          commands -= CMD_TRUNCATE;
        }
      else if (commands != CMD_NONE)
        {
          WARNING ("Unknown command[s] " << commands << " left unprocessed.");
          commands = CMD_NONE;
        }
    }

  if (cleaner != NULL)
    delete cleaner;

  process.stop ();

  LocalNetworks::destroy ();
  TemplateList::destroy ();
  SAMSUserList::destroy ();
  Proxy::destroy ();
  Logger::destroy ();
  SamsConfig::destroy ();
}
